### ISSUE TITLE
Support MemSQL

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#2935]
MemSQL is "A Fully-Managed, On-Demand, & Elastic Cloud Database" that is compatible with MySQL.

**Comment by janpio** [comment]
Currently this happens on `prisma introspect` or `prisma migrate save --experimental`:

```
C:\Users\Jan\Documents\throwaway\memsql>prisma introspect
Environment variables loaded from prisma\.env

Introspecting based on datasource defined in prisma\schema.prisma …
  IntrospectionEngine:rpc starting introspection engine with binary: C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\introspection-engine-windows.exe +0ms
  IntrospectionEngine:rpc SENDING RPC CALL {"id":1,"jsonrpc":"2.0","method":"introspect","params":[{"schema":"// This is your Prisma schema file,\n// learn more about it in the docs: https://pris.ly/d/prisma-schema\n\ndatasource db {\n  provider = \"mysql\"\n  url      = env(\"DATABASE_URL\")\n}\n\ngenerator client {\n  provider = \"prisma-client-js\"\n}\n\nmodel User {\n  id    String  @default(cuid()) @id\n  email String  @unique\n  name  String?\n}"}]} +7ms
  IntrospectionEngine:rpc {
  IntrospectionEngine:rpc   jsonrpc: '2.0',
  IntrospectionEngine:rpc   error: {
  IntrospectionEngine:rpc     code: 4466,
  IntrospectionEngine:rpc     message: 'An error happened. Check the data field for details.',
  IntrospectionEngine:rpc     data: {
  IntrospectionEngine:rpc       is_panic: false,
  IntrospectionEngine:rpc       message: "Error in connector: Error querying the database: Error querying the database: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'Audit logging with binary protocol' is not supported by MemSQL.'",
  IntrospectionEngine:rpc       backtrace: null
  IntrospectionEngine:rpc     }
  IntrospectionEngine:rpc   },
  IntrospectionEngine:rpc   id: 1
  IntrospectionEngine:rpc } +950ms
Error: Error: Error in connector: Error querying the database: Error querying the database: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'Audit logging with binary protocol' is not supported by MemSQL.'

    at IntrospectionEngine.<anonymous> (C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1995969)
    at Generator.next (<anonymous>)
    at C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1991014
    at new Promise (<anonymous>)
    at n (C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1990762)
    at Object.<anonymous> (C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1995175)
    at IntrospectionEngine.handleResponse (C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1993052)
    at LineStream.<anonymous> (C:\ProgramData\nvm\v12.6.0\node_modules\@prisma\cli\build\index.js:2:1994813)
    at LineStream.emit (events.js:203:13)
    at addChunk (_stream_readable.js:294:12)
```

Seems it does not like prepared statements in the default configuration: https://docs.memsql.com/v7.1/concepts/prepared-statements/

**Comment by braposo** [comment]
Hey @janpio I was wondering if this has had any updates since last year?

I've recently joined SingleStore and was trying out the latest release of Prisma Migrate. I was able to get a bit further than you but still getting errors from the introspection engine.

```
braposo@braposo-1:~/Projects/hello-prisma$ npx prisma migrate dev --name init --create-only
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "test" at "172.21.0.3:30141"

Error: Database error: Error accessing result set, column not found: datetime_precision
   0: sql_migration_connector::flavour::mysql::sql_schema_from_migration_history
             at migration-engine/connectors/sql-migration-connector/src/flavour/mysql.rs:270
   1: sql_migration_connector::sql_database_migration_inferrer::calculate_drift
             at migration-engine/connectors/sql-migration-connector/src/sql_database_migration_inferrer.rs:57
   2: migration_core::api::DevDiagnostic
             at migration-engine/core/src/api.rs:97
```

The `columns` table in the `information_schema` database doesn't contain the required `datetime_precision` column so I was wondering if there's a way to disable that check somehow?

I found this issue from another project talking about a similar problem: https://github.com/go-gorm/mysql/issues/13

Please let me know if there's anything I can do to help getting Prisma to work with SingleStore!

**Comment by braposo** [comment]
I've also tested with `@prisma/client` and it's actually working for the reading (find, aggregate, count, etc) actions, but there's some syntax issue for the writing actions (create, update, delete).

```
 Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(ServerError { code: 1064, message: "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near \'? LIMIT ? OFFSET ?\' at line 1", state: "42000" })) })
    at cb (/home/braposo/Projects/hello-prisma/node_modules/@prisma/client/runtime/index.js:34778:17)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

Is there a way to see what's the actual SQL query that is being sent? I've tried enabling debugging mode but it doesn't print that.

Otherwise it's actually working really nicely, which is very exciting!

![image](https://user-images.githubusercontent.com/38172/111506952-11228b80-8742-11eb-8556-ae1e19b9f739.png)

Do you want me to create a separte issue for this or is it fine to talk about this here as well?

**Comment by janpio** [comment]
> The `columns` table in the `information_schema` database doesn't contain the required `datetime_precision` column so I was wondering if there's a way to disable that check somehow?

We would need to adapt our Migration and Introspection Implementation to skip that step for SingleStore - which might not be trivial.

In general 2 things would be helpful:
1. Is there a Docker image that we can easily integrate into our test suite?
2. Is there a reliable way to "detect" and identify a SingleStore instance vs. a default MySQL?

> Is there a way to see what's the actual SQL query that is being sent? I've tried enabling debugging mode but it doesn't print that.

https://www.prisma.io/docs/concepts/components/prisma-client/debugging and https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging should give you what you are looking for I think to see the queries that are being executed under the hood and pinpoint the one giving you the error.

Generally SingleStore is not a priority for us, as our customers and users are asking for other things. But I am happy to help you out answering questions and possibly also seeing about getting SingelStore into our CI setup. If we find deeper problems (like the first point about `datetime_precision`) I might not be able to guarantee Engineering capacity to fix those any time soon - but let's see how far we get until that is the case!

**Comment by braposo** [comment]
hey @janpio! First of all thank you so much for the quick reply and I'm sorry for not getting back to you earlier but it was hackathon at SingleStore and I couldn't get all the answers for you in time.

> Not out of the box (I think this is coming from [prisma/prisma-engines@`e106258`/libs/sql-schema-describer/src/mysql.rs#L228-L252](https://github.com/prisma/prisma-engines/blob/e106258e8db7c70b2cc7b914a0bc4e78b1c35d1c/libs/sql-schema-describer/src/mysql.rs#L228-L252)). We would need to adapt our implementation there to skip that step for SingleStore, adapt the query or similar - which might not be trivial.

No problem, I wasn't really expecting a workaround from your side, was more to understand what you were querying for and your answer is very helpful! I'll be working with some people from the engine team to see if I can unblock the remaining integration problems from our side.

> In general 2 things would be helpful:
> 
> 1. Is there a Docker image that we can easily integrate into our test suite? (a la [https://github.com/prisma/prisma-engines/blob/master/docker-compose.yml](https://github.com/prisma/prisma-engines/blob/master/docker-compose.yml?rgh-link-date=2021-03-17T21%3A59%3A53Z))
> 2. Is there a reliable way to "detect" and identify a SingleStore instance vs. a default MySQL?

1. Yes there is! We publish a [memsql-in-a-box docker image](https://hub.docker.com/r/memsql/cluster-in-a-box/tags?page=1&ordering=last_updated) that you can use in your CI. For example:

```
singlestore:
    image: 'memsql/cluster-in-a-box:centos-7.3.6-ae5d1857b8-3.2.5-1.11.0'
    restart: always
    environment:
      ROOT_PASSWORD: "password"
      LICENSE_KEY: $LICENSE_KEY 
      START_AFTER_INIT: true
    ports:
      - "3306:3306"
```
To get a free license key you can signup at http://portal.singlestore.com/ but please let me know if you have any trouble with that.

2. An easy way to check is by running `select @@memsql_version` query which will only work for SingleStore DB

> [prisma.io/docs/concepts/components/prisma-client/debugging](https://www.prisma.io/docs/concepts/components/prisma-client/debugging) and [prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging) should give you what you are looking for I think to see the queries that are being executed under the hood and pinpoint the one giving you the error.

Thanks! I was able to get some additional logging from your suggestions and was able to reproduce the issue using the `mysql2` library so this isn't exclusive to your package. Do you know which mysql connector are you using in the background? Is it a custom one?

> Generally SingleStore is not a priority for us, as our customers and users are asking for other things. But I am happy to help you out answering questions and possibly also seeing about getting SingelStore into our CI setup. If we find deeper problems (like the first point about `datetime_precision`) I might not be able to guarantee Engineering capacity to fix those any time soon - but let's see how far we get until that is the case!

Totally understand that and I believe we'll be able to ge the integration working from our side, your guidance would be great! This isn't something we're working on actively either, it's mostly a side/pet project of mine, but hopefully I can dedicate some time to it!

**Comment by janpio** [comment]
Nice, thanks for your reply and answers. I will use the Docker container to set up a small side project of mine as well to just run our test suites against it and see what incompatibilities fall out of that.

**Comment by tomhoule** [comment]
> Do you know which mysql connector are you using in the background? Is it a custom one?

We are using the [mysql_async crate](https://crates.io/crates/mysql_async) to connect to MySQL.

**Comment by janpio** [comment]
Good that I have a day off today @braposo, so I could already play around a bit to get this into our CI.

First bump in getting the test for our Introspection Engine, which introspects the database schema and creates a Prisma Schema from it (and is used in most of our other tests to confirm results of the Migration Engine, and database setup for Query Engine tests), to run was that we rely on the `mysql` database being present to connect to the database server. Fixed via a small query after starting the server: `- run: mysql -h 127.0.0.1 -P 3306 -u root -pprisma -e "CREATE DATABASE mysql;" --loose-default-auth=mysql_native_password` (Also documented for your users at https://www.singlestore.com/forum/t/mysqldb-exceptions-operationalerror-2012-error-in-server-handshake/2242/5?u=janpio.prisma)

Second one was that the database just needed a bit to start - so the above only worked after waiting a bit.

Afterwards I hit multiple different problems:

- `Error querying the database: Error accessing result set, column not found: datetime_precision` you mentioned above already.
- Then I quickly get a space error for some reason: 
  ```
  thread 'commenting_out::a_table_without_required_uniques::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1772, message: "CREATE PARTITION DATABASE for operation `CREATE DATABASE` failed with error 1883:Leaf Error (127.0.0.1:3307): Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.", state: "HY000" })), original_code: Some("1772"), original_message: Some("CREATE PARTITION DATABASE for operation `CREATE DATABASE` failed with error 1883:Leaf Error (127.0.0.1:3307): Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.") }', /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/test_api.rs:39:18
  ```
  or
  ```
  thread 'relations::relations_should_avoid_name_clashes_2::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1883, message: "Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.", state: "HY000" })), original_code: Some("1883"), original_message: Some("Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.") }', /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/test_api.rs:39:18
  ```
  Source: https://github.com/prisma/prisma-engines/pull/1802/checks?check_run_id=2245842332
  Some googling told me this was a error message from the Singlestore server (e.g. https://www.singlestore.com/forum/t/cluster-in-a-box-minimal-disk-space-error/2503), but my tries to fix it by adding a volume to the Docker Container etc did not work - I could not find a copy pastable example.
  
Can you help with that maybe? Current code: https://github.com/prisma/prisma-engines/pull/1802/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R4

**Comment by janpio** [comment]
No fun in waiting: I quickly modified the query that uses `datetime_precision` to always return `NULL` instead (https://github.com/prisma/prisma-engines/pull/1802/files#diff-9996eb1f4322b87712082043032bbd3ca8a0e8ccde7ecf19d80fefd7968c03b9L244), and that unblocked these two tests to make them succeed 💯 

So if we can fix this space error, we might get many more tests succeeding as well 👍

**Comment by braposo** [comment]
oh wow this is great @janpio! Let me reach out to the people that might be able to figure out what's going on with docker and hopefully we can get over that hurdle as well! 

Thank you so much for your help, this is very exciting! (at least for me :sweat_smile:!)

**Comment by FranciscoMSM** [comment]
Hello @janpio! Thank you for your effort on this one! :smile: 

That error indicates you don't have enough disk space to create the database.
Can you please run `df -h` in the container to understand the amount of free disk space we're working on with?

Thank you!

**Comment by janpio** [comment]
This is a GH Actions run on an Ubuntu container - usually there are around ~15GB free space available there. This is the https://github.com/prisma/prisma-engines/blob/janpio-patch-7/docker-compose.yml file I am running `docker-compose up -d` on @FranciscoMSM. We probably just have to give some of that available space to the container.

**Comment by janpio** [comment]
I sprinkled a bunch of commands into that workflow:

```
      - run: df -h
      - run: docker-compose run mysql-8-0 df -h
```

Results will be at https://github.com/prisma/prisma-engines/pull/1802/checks?check_run_id=2247005960 in a bit.

**Comment by braposo** [comment]
Could it be related to this comment? https://www.singlestore.com/forum/t/cluster-in-a-box-minimal-disk-space-error/2503/11

would something like this work?

```
volumes:
      - /memsql:/var/lib/memsql:rw
```

I also noticed that in other examples it's using `tmpfs`, have you tried:
```
tmpfs: /var/lib/memsql
```

sorry if I can't be more helpful at this point but hopefully @FranciscoMSM can come with better suggestions if this doesn't work!

**Comment by janpio** [comment]
I do not know the difference between `volumes`, `tmpfs`, and even the syntax of `volumes` is not totally clear to me and my google is failing me.

I tried with `tmpfs: /var/lib/memsql` now and that makes this line appear in `df -h` output on the docker container:
```
tmpfs           3.4G     0  3.4G   0% /var/lib/memsql
```

But then I can not connect to the database:
```
> Run mysql -h 127.0.0.1 -P 3307 -u root -pprisma -e "SHOW DATABASES;" --loose-default-auth=mysql_native_password
ERROR 2003 (HY000): Can't connect to MySQL server on '127.0.0.1' (111)
```

So something I am doing wrong somewhere.

**Comment by janpio** [comment]
Ok, after a lot of fiddling after dinner I have many more tests running 💯 

I am now using a GH Action to free up more space on the machine, and then make that space available to `/var/lib/memsql` inside Docker: 65G

```
Filesystem                   Size  Used Avail Use% Mounted on
/dev/mapper/buildvg-buildlv   65G   80M   65G   1% /var/lib/memsql
```

That means I can now actually _run_ a lot more tests, and get lots of real, actual errors:


```
---- commenting_out::a_table_without_fully_required_compound_unique::mysql_8 stdout ----
thread 'commenting_out::a_table_without_fully_required_compound_unique::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1896): Cannot declare primary key or unique key named: 'sqlite_autoindex_Post_1' when using an empty shard key'

Location:
    /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/lib.rs:58:47', introspection-engine/introspection-engine-tests/tests/commenting_out/mod.rs:127:1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- commenting_out::a_table_without_required_uniques::mysql_8 stdout ----
thread 'commenting_out::a_table_without_required_uniques::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1896): Cannot declare primary key or unique key named: 'opt_unique' when using an empty shard key'

Location:
    /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/lib.rs:58:47', introspection-engine/introspection-engine-tests/tests/commenting_out/mod.rs:101:1

---- commenting_out::a_table_without_uniques_should_ignore::mysql_8 stdout ----
thread 'commenting_out::a_table_without_uniques_should_ignore::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'FOREIGN (non-SHARD) key to a sharded table' is not supported by MemSQL.'

Location:
    /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/lib.rs:58:47', introspection-engine/introspection-engine-tests/tests/commenting_out/mod.rs:7:1

---- enums::a_table_enums_should_return_alphabetically_even_when_in_different_order::mysql_8 stdout ----
thread 'enums::a_table_enums_should_return_alphabetically_even_when_in_different_order::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1778, message: "Distributed MemSQL requires the length of the database name to be at most 60 characters.", state: "HY000" })), original_code: Some("1778"), original_message: Some("Distributed MemSQL requires the length of the database name to be at most 60 characters.") }', /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/test_api.rs:39:18

---- commenting_out::remapping_field_names_to_empty::mysql_8 stdout ----
thread 'commenting_out::remapping_field_names_to_empty::mysql_8' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/remapping_field_names_to_empty?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel User {\n  last Int @id @default(autoincrement())\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/remapping_field_names_to_empty?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel User {\n  last BigInt @id @default(autoincrement())\n}\n"

', introspection-engine/introspection-engine-tests/src/test_api.rs:220:9

---- enums::a_table_with_enum_default_values::mysql_8 stdout ----
thread 'enums::a_table_with_enum_default_values::mysql_8' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enum_default_values?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id    Int     @id @default(autoincrement())\n  color Book_color @default(black)\n}\n\nenum Book_color {\n  black\n  white\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enum_default_values?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id    BigInt     @id @default(autoincrement())\n  color Book_color @default(black)\n}\n\nenum Book_color {\n  black\n  white\n}\n"

', introspection-engine/introspection-engine-tests/src/test_api.rs:220:9

---- enums::a_table_with_an_enum_default_value_that_is_an_empty_string::mysql_8 stdout ----
thread 'enums::a_table_with_an_enum_default_value_that_is_an_empty_string::mysql_8' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_an_enum_default_value_that_is_an_empty_string?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id    Int     @id @default(autoincrement())\n  color Book_color @default(EMPTY_ENUM_VALUE)\n}\n\nenum Book_color {\n  black\n  EMPTY_ENUM_VALUE @map(\"\")\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_an_enum_default_value_that_is_an_empty_string?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id    BigInt     @id @default(autoincrement())\n  color Book_color @default(EMPTY_ENUM_VALUE)\n}\n\nenum Book_color {\n  black\n  EMPTY_ENUM_VALUE @map(\"\")\n}\n"

', introspection-engine/introspection-engine-tests/src/test_api.rs:220:9

---- enums::a_table_with_enums::mysql_8 stdout ----
thread 'enums::a_table_with_enums::mysql_8' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enums?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id     Int      @id @default(autoincrement())\n  color  Book_color\n  color2 Book_color2\n}\n\nenum Book_color {\n  black\n  white\n}\n\nenum Book_color2 {\n  black2\n  white2\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enums?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel Book {\n  id     BigInt      @id @default(autoincrement())\n  color  Book_color\n  color2 Book_color2\n}\n\nenum Book_color {\n  black\n  white\n}\n\nenum Book_color2 {\n  black2\n  white2\n}\n"

', introspection-engine/introspection-engine-tests/src/test_api.rs:220:9

---- enums::a_table_with_enum_default_values_that_look_like_booleans::mysql_8 stdout ----
thread 'enums::a_table_with_enum_default_values_that_look_like_booleans::mysql_8' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enum_default_values_that_look_like_booleans?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel News {\n  id        Int         @id @default(autoincrement())\n  confirmed News_confirmed @default(true)\n}\n\nenum News_confirmed {\n  true\n  false\n  rumor\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/a_table_with_enum_default_values_that_look_like_booleans?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel News {\n  id        BigInt         @id @default(autoincrement())\n  confirmed News_confirmed @default(true)\n}\n\nenum News_confirmed {\n  true\n  false\n  rumor\n}\n"

', introspection-engine/introspection-engine-tests/src/test_api.rs:220:9

---- identify_version::introspect_mysql_non_prisma::mysql_8 stdout ----
thread 'identify_version::introspect_mysql_non_prisma::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR 42000 (1064): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'point)' at line 1'

Location:
    /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/lib.rs:58:47', introspection-engine/introspection-engine-tests/tests/identify_version/mod.rs:186:1

---- identify_version::introspect_mysql_prisma_1::mysql_8 stdout ----
thread 'identify_version::introspect_mysql_prisma_1::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'DATETIME type with precision other than 0 or 6' is not supported by MemSQL.'

Location:
    /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/lib.rs:58:47', introspection-engine/introspection-engine-tests/tests/identify_version/mod.rs:202:1

---- identify_version::introspect_mysql_prisma_1_1::mysql_8 stdout ----
thread 'identify_version::introspect_mysql_prisma_1_1::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1842, message: "Database name (introspect_mysql_prisma_1_1) conflicts with an existing partition database name (partition databases are formatted as database_#, where # is a number)", state: "HY000" })), original_code: Some("1842"), original_message: Some("Database name (introspect_mysql_prisma_1_1) conflicts with an existing partition database name (partition databases are formatted as database_#, where # is a number)") }', /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/test_api.rs:39:18

---- identify_version::introspect_mysql_prisma2::mysql_8 stdout ----
thread 'identify_version::introspect_mysql_prisma2::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1772, message: "CREATE PARTITION DATABASE for operation `CREATE DATABASE` failed with error 1883:Leaf Error (127.0.0.1:3307): Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.", state: "HY000" })), original_code: Some("1772"), original_message: Some("CREATE PARTITION DATABASE for operation `CREATE DATABASE` failed with error 1883:Leaf Error (127.0.0.1:3307): Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.") }', /home/runner/work/prisma-engines/prisma-engines/introspection-engine/introspection-engine-tests/src/test_api.rs:39:18
```

After that we are back to many, many of the same space errors unfortunately :( 

And indeed:

```
/dev/mapper/buildvg-buildlv   65G   62G  3.0G  96% /var/lib/memsql
```

Not the 100MB the error message mentions, but 62G are now filled with _something_.

Can you help me make the Docker container use less space? Or at least understand what is taking up that space @braposo?

**Comment by janpio** [comment]
Back to the list of actual errors:

First three are about keys and empty shard key and sharded table:

- thread 'commenting_out::a_table_without_fully_required_compound_unique::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1896): Cannot declare primary key or unique key named: 'sqlite_autoindex_Post_1' when using an empty shard key'
- thread 'commenting_out::a_table_without_required_uniques::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1896): Cannot declare primary key or unique key named: 'opt_unique' when using an empty shard key'
- thread 'commenting_out::a_table_without_uniques_should_ignore::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'FOREIGN (non-SHARD) key to a sharded table' is not supported by MemSQL.'
- 
No idea what any of this means.

---

One about using a to long table name:
- thread 'enums::a_table_enums_should_return_alphabetically_even_when_in_different_order::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1778, message: "Distributed MemSQL requires the length of the database name to be at most 60 characters.", state: "HY000" })), original_code: Some("1778"), original_message: Some("Distributed MemSQL requires the length of the database name to be at most 60 characters.") }'
- 
That one is just configuration so should be easy to fix.

---

Multiple about a `Int`/`BigInt` mismatch in the returned schema:
- thread 'commenting_out::remapping_field_names_to_empty::mysql_8' panicked at 'assertion failed: `(left == right)`
```diff
Diff < left / right > :
<"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/remapping_field_names_to_empty?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel User {\n  last Int @id @default(autoincrement())\n}\n"
>"datasource db {\n  provider = \"mysql\"\n  url      = \"***127.0.0.1:3307/remapping_field_names_to_empty?connect_timeout=20&socket_timeout=60\"\n}\n\nmodel User {\n  last BigInt @id @default(autoincrement())\n}\n"
```

My guess would be that Signelstore somehow creates a different column as we tell it to because it is smart in some way? Or the information about that difference is represented differently maybe?

---

A MySQL syntax error:
- thread 'identify_version::introspect_mysql_non_prisma::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR 42000 (1064): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'point)' at line 1'

Possibly something not supported by SingleStore.

--- 

Something about datetime precision:
- thread 'identify_version::introspect_mysql_prisma_1::mysql_8' panicked at 'The test function returned an Err.: Error querying the database: Server error: `ERROR HY000 (1706): Feature 'DATETIME type with precision other than 0 or 6' is not supported by MemSQL.'

I bet this is a side effect of the hack that I added to unblock the missing  `datetime_precision`

---

And finally something about a name clash:
- thread 'identify_version::introspect_mysql_prisma_1_1::mysql_8' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(Server(ServerError { code: 1842, message: "Database name (introspect_mysql_prisma_1_1) conflicts with an existing partition database name (partition databases are formatted as database_#, where # is a number)", state: "HY000" })), original_code: Some("1842"), original_message: Some("Database name (introspect_mysql_prisma_1_1) conflicts with an existing partition database name (partition databases are formatted as database_#, where # is a number)") }'

I guess Singlestore has rules about what you can call the tables, that conflicts with what we generate automatically for the test.

**Comment by braposo** [comment]
This is some incredible work @janpio! Thank you so much for doing this and for taking the time to get this a bit further!

I'm going to take this back to the team and try to find the answers to your questions. I'll let you know when I have something to share.

In the meantime, is there a way for us to test SingleStore against your test suite outside of your CI? It's probably faster if we can run the tests locally as we're making the changes to the engine.

Thanks again, this is really amazing!

**Comment by janpio** [comment]
The PR linked above shows pretty much that in its GH Actions configuration.
TLDR is: Check out https://github.com/prisma/prisma-engines, install the 1.48.0 Rust toolchain, make a DB run at `localhost:3007` (plus create an empty `mysql` database) and then execute `cargo test mysql_8` in `introspection-engine/introspection-engine-tests`. (As Singlestore is MySQL compatible (kinda), this "reuses" our MySQL (8.0) test suite (which usually runs against a MySQL DB at 3007). This avoids us having to add all the usual boilerplate to add a new database and the reason why I could do so as a small side project.)

I could continue with Migration and Query Engine tests here now, but I would probably as quickly hit the memory problem again. And as both these test suites are considerably bigger than the Introspection Engine one, the % of tests we can run will probably be a lot lower :/ So looking forward to the feedback on how to tweak the container configuration to use less space.

**Comment by janpio** [comment]
(I ran our Migration Engine and Query Engine tests on SingleStore as well - and as expected `Leaf Error (127.0.0.1:3307): Available disk space is below the value of \'minimal_disk_space\' global variable (100 MB). This query cannot be executed.` prevents most to all tests being run for now.)

**Comment by janpio** [comment]
Seems some logging is going crazy:

```
root@fv-az186-340:/home/runner/work/prisma-engines/prisma-engines/memsql/8b0b24b6-1868-4160-8baf-e6a07d34ee89/data/logs# ls -l
...
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_1_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_1_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_2_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_2_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_3_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_3_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_4_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_4_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_5_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_5_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_6_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_6_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_7_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 introspect_mysql_non_prisma_empty_7_log_v1_65536
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:54 introspect_mysql_non_prisma_empty_log_v1_0
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:53 introspect_mysql_non_prisma_empty_log_v1_16384
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:54 introspect_mysql_non_prisma_log_v1_0
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:53 introspect_mysql_non_prisma_log_v1_16384
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_0_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_0_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_1_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_1_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_2_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_2_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_3_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_3_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_4_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_4_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_5_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_5_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_6_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_6_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_7_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:54 introspect_mysql_prisma2_7_log_v1_65536
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:54 introspect_mysql_prisma2_log_v1_0
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:54 introspect_mysql_prisma2_log_v1_16384
-rw------- 1 systemd-coredump 998  16777216 Apr  4 14:54 memsql_log_v1_0
-rw------- 1 systemd-coredump 998  16777216 Apr  4 14:47 memsql_log_v1_4096
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_0_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_0_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_1_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_1_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_2_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_2_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_3_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_3_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_4_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_4_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_5_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_5_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_6_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_6_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_7_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:48 mysql_7_log_v1_65536
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:48 mysql_log_v1_0
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:48 mysql_log_v1_16384
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_0_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_0_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_1_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_1_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_2_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_2_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_3_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_3_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_4_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_4_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_5_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_5_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_6_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_6_log_v1_65536
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_7_log_v1_0
-rw------- 1 systemd-coredump 998 268435456 Apr  4 14:53 remapping_field_names_to_empty_7_log_v1_65536
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:54 remapping_field_names_to_empty_log_v1_0
-rw------- 1 systemd-coredump 998  67108864 Apr  4 14:53 remapping_field_names_to_empty_log_v1_16384
root@fv-az186-340:/home/runner/work/prisma-engines/prisma-engines/memsql/8b0b24b6-1868-4160-8baf-e6a07d34ee89/data/logs#
```

58G of logging that is super verbose, but not very helpful. Any way I can disable that in the container @braposo @FranciscoMSM? That might already be enough to get the tests running.

**Comment by janpio** [comment]
Ok, those seem to be snapshots of the databases. No idea why they are so huge. I could not find a way to disable snapshots either. Right now it seems as creating many databases is incompatible with this Docker container :/

**Comment by janpio** [comment]
> Now I'm trying to get some help on how to set it as part of the container in a box image so I'll get back to you when I get that information.

I _think_ this is of the same class as `core_file` which I played with already - so I think I might know already. (Link to _really_ terrible code, where I messed with those thinking it would help: https://github.com/prisma/prisma-engines/pull/1802/files#diff-32349d539eeee14551827f0971cfb90f048fa8bceb05904631f4838c530fa865R51-R65)

**Comment by salemhilal** [comment]
Hey! I've run into the same error as you did initially (`Error accessing result set, column not found: datetime_precision`) when trying to do my very first `prisma db pull` on an existing DB we have. I wanted to check in and see if there are any updates in this area, or if I can help debug. I'm new to both singlestore and prisma, so you may have to be a little patient with me.

**Comment by janpio** [comment]
No, that error message indicates that one of the fundamental columns that we need for `db pull` and that usually exists in MySQL, does not in SingleStore. We would need to adapt our code (after finding an alternative) to make this work.

**Comment by MohannadBahaa** [comment]
Hello @janpio 

using Prisma with Singlestore is quite hard so far !
when it comes to the DDL, push changes, indices ...etc !

is there any plan sooner to support Singlestore

**Comment by janpio** [comment]
No, we are not working on improving support for SingleStore right now. We have too many other things to do first (see other issues, roadmap and releases) before we will get to add additional connectors/databases that do not work out of the box.

**Comment by MohannadBahaa** [comment]
@janpio make sense thank you

**Comment by xpluscal** [comment]
@MohannadBahaa did you move to another solution or did you find a way to get SingleStore to work with Prisma?

**Comment by MohannadBahaa** [comment]
@xpluscal unfortunately, I have moved with Sequelize check this [start-with-singlestore-node-orm](https://github.com/singlestore-labs/start-with-singlestore-node-orm)

**Comment by Zonalds** [comment]
@MohannadBahaa, what was the experience using Sequelize on Singlestore? Had any issue with Sequelize using Singlestore?

**Comment by MohannadBahaa** [comment]
Experience: Consuming data from SQS then using Sequelize for copying to Singlestore.

Cons: Managing the connection pool was a challenge. This was particularly the case when dealing with a high volume of data - around 1 million records per hour.
